File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    int n, m;
    int[] parent;

    Main(int n, int m) {
        this.n = n;
        this.m = m;
        parent = new int[n + 1];
        for (int i = 0; i < n + 1; i++) {
            parent[i] = i;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        Main main = new Main(n, m);

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int operation = Integer.parseInt(st.nextToken());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            if (operation == 0) {
                main.union(a, b);
            } else if (operation == 1) {
                boolean isSame = main.check(a, b);
                sb.append((isSame ? "YES" : "NO") + "\n");
            }
        }
        System.out.print(sb);
    }

    public int find(int x) {
        if (x == parent[x]) {
            return x;
        }
        return parent[x] = find(parent[x]);
    }

    public boolean check(int a, int b) {
        return find(a) == find(b);
    }

    public void union(int a, int b) {
        int x = find(a);
        int y = find(b);
        if (x > y) {
            parent[x] = y;
        } else {
            parent[y] = x;
        }
    }
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)


def find(n):
    # if n == parents[n]:
    #     return n
    # parents[n] = find(parents[n])
    # return parents[n]
    if n != parents[n]:
        parents[n] = find(parents[n])
    return parents[n]


def union(n1, n2):
    num1 = find(n1)
    num2 = find(n2)
    if num1 > num2:
        parents[num1] = num2
    else:
        parents[num2] = num1


N, M = map(int, input().split())
parents = [i for i in range(N+1)]
for _ in range(M):
    op, n1, n2 = map(int, input().split())
    if op == 0:
        union(n1, n2)
    elif op == 1:
        p1 = find(n1)
        p2 = find(n2)
        if p1 == p2:
            print("YES")
        else:
            print("NO")

